pilot/pkg/networking/core/v1alpha3/httproute.go

```diff
@@ -199,11 +210,25 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundHTTPRouteConfig(env *m
 		vHostPortMap[virtualHostWrapper.Port] = append(vHostPortMap[virtualHostWrapper.Port], virtualHosts...)
 	}
 
-	var virtualHosts []*route.VirtualHost
+	var tmpVirtualHosts []*route.VirtualHost
 	if listenerPort == 0 {
-		virtualHosts = mergeAllVirtualHosts(vHostPortMap)
+		tmpVirtualHosts = mergeAllVirtualHosts(vHostPortMap)
+	} else {
+		tmpVirtualHosts = vHostPortMap[listenerPort]
+	}
+
+	var virtualHosts []*route.VirtualHost
+	if useSniffing {
+		for _, vh := range tmpVirtualHosts {
+			for _, domain := range vh.Domains {
+				if domain == routeName {
+					virtualHosts = append(virtualHosts, vh)
+					break
+				}
+			}
+		}
```

Comment:
you also need a fallback logic here. if virtualHosts is empty, then you are better off shipping the entire tmpVirtualHosts. This fallback is needed if you are doing that bindIP thing in listener https://github.com/istio/istio/pull/16876/files#diff-86fb73d7dd150e47328ead76ceca0992R1049 - I am not sure why its needed though